### PR TITLE
Issue #5176: adjacent static import groups ones became impossible in 8.3

### PR DESCRIPTION
**Pull request by djydewang**
Issue #5176 and Issue #5253

### CODE REVIEW

**Comment by djydewang** [comment]
There is a rough [groovy diff report](https://djydewang.github.io/).

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 1]
```diff
@@ -137,6 +137,51 @@ public void testStaticImportSeparated() throws Exception {
...
+        checkConfig.addAttribute("separated", "true");
+        checkConfig.addAttribute("ordered", "true");
+        checkConfig.addAttribute("option", "top");
+        checkConfig.addAttribute("sortStaticImportsAlphabetically", "true");
+        final String[] expected = {
+            "4: " + getCheckMessage(MSG_ORDERING,
+                "javax.xml.transform.TransformerFactory.newInstance"),
+            "5: " + getCheckMessage(MSG_ORDERING, "java.lang.Math.cos"),
+            "6: " + getCheckMessage(MSG_ORDERING, "java.lang.Math.abs"),
+            };
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: ident this line left 4 spaces. It should align with the `final`.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 2]
```diff
@@ -137,6 +137,51 @@ public void testStaticImportSeparated() throws Exception {
...
+    public void testsortStaticImportsAlphabeticallyFalse() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(ImportOrderCheck.class);
+        checkConfig.addAttribute("groups", " java, javax, org");
+        checkConfig.addAttribute("separated", "true");
+        checkConfig.addAttribute("ordered", "true");
+        checkConfig.addAttribute("option", "top");
+        checkConfig.addAttribute("sortStaticImportsAlphabetically", "false");
+        final String[] expected = CommonUtils.EMPTY_STRING_ARRAY;
+
+        verify(checkConfig, getPath("InputImportOrdersortStaticImportsAlphabetically.java"),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: `sort` => `Sort`
Casing should be camel case.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 3]
```diff
@@ -137,6 +137,51 @@ public void testStaticImportSeparated() throws Exception {
...
+        checkConfig.addAttribute("separated", "true");
+        checkConfig.addAttribute("ordered", "true");
+        checkConfig.addAttribute("option", "bottom");
+        final String[] expected = CommonUtils.EMPTY_STRING_ARRAY;
+
+        verify(checkConfig, getPath("InputImportOrderNoGapBetweenStaticImports.java"), expected);
+    }
+
+    @Test
+    public void testsortStaticImportsAlphabeticallyFalse() throws Exception {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: same for test name.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 4]
```diff
@@ -137,6 +137,51 @@ public void testStaticImportSeparated() throws Exception {
...
+        checkConfig.addAttribute("option", "top");
+        checkConfig.addAttribute("sortStaticImportsAlphabetically", "false");
+        final String[] expected = CommonUtils.EMPTY_STRING_ARRAY;
+
+        verify(checkConfig, getPath("InputImportOrdersortStaticImportsAlphabetically.java"),
+            expected);
+    }
+
+    @Test
+    public void testsortStaticImportsAlphabeticallyTrue() throws Exception {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: same for test name.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 2 above)
Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 3 above)
Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 2 above)
Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 4 above)
Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 1 above)
Comment (djydewang) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: Done.

**Comment by djydewang** [comment]
[pit report](https://djydewang.github.io/pit-reports/index.html)
One commit for one mutation coverage.

**Comment by djydewang** [comment]
There is the [pit report](https://djydewang.github.io/pit-add-repetition-report/index.html) after adding repetition import test.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 5]
```diff
@@ -546,20 +546,20 @@ private boolean isWrongOrder(String name, boolean isStatic) {
...
             }
             else {
-                result = compare(lastImport, name, caseSensitive) > 0;
+                result = compare(lastImport, name, caseSensitive) >= 0;
             }
         }
         else {
             // out of lexicographic order
-            result = compare(lastImport, name, caseSensitive) > 0;
+            result = compare(lastImport, name, caseSensitive) >= 0;
```
Comment (djydewang) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: Should we throw other warning message for repetitive imports? Or just import ordering message.

**Comment by romani** [comment]
> Please review it and give me some advice for next commit.

Do you mean on how to cover lines ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: report only on import ordering.
Fact that import is duplicated is not this check business.

Comment (djydewang) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: > Fact that import is duplicated is not this check business.

Right, maybe we can add new feature to detect duplicated import in the future.But, I can only cover the red lines by this method now.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: in this PR please focus on passing CIs

**Comment by djydewang** [comment]
@romani 
> We know what this method suppose to do, we need to change his logic to simplify it and remove unnecessary expressions.

His logic is all correct.
And I just add a test case for it and avoid using reflection.
All red lines are covered now.

**Comment by romani** [comment]
@djydewang , thanks a lot.

please squash 1-2 commit as 2nd commit change of code that was introduced in 1st. Lets presume that you did all correctly in one commit.

3rd commit is ok to keep separate, as it just covering unrelated to issue part of code.
As a prove that you covered all lines please remove exlclude for ImportOrder at https://github.com/checkstyle/checkstyle/blob/master/shippable.yml#L64 , to let CI enforce 100% coverages for any future updates. As it is incredible achievements that was done by you.

**Comment by djydewang** [comment]
@romani 
I rechecked the pitest report since CIs were failed and find that one of  red lines covered  by memery error. 
So, I try again to cover it.
original code:
`length > bestLength || length == bestLength && matcher.start() < bestPos
`
His idea is that matching length have high priority.

But I think that matching position of start have high priority.
`matcher.start() < bestPos || matcher.start() == bestPos && matcher.end() > bestEnd.`

**Comment by djydewang** [comment]
It passed `mvn clean veerify` locally.
I don't konw why appveyor is failing on code coverage.

**Comment by romani** [comment]
> I don't konw why appveyor is failing on code coverage.

sorry it was unexpected mess - #5283.
Should be fixed.

> Run 5581 status is FAILED. 

your code did not cause it, I did one more fix in scope of #5283 , your shippable was fine previous time, so I will fix you PR as I finish last review. Sorry for mess with CIs.

File: src/xdocs/config_imports.xml [hunk 6]
```diff
@@ -1291,7 +1291,8 @@ import java.util.stream.IntStream;
           </tr>
           <tr>
             <td>ordered</td>
-            <td>whether imports within group should be sorted</td>
+            <td>whether imports within group should be sorted
+              (It don't affect soring for static imports.)</td>
```
Comment (rnveach) [inline src/xdocs/config_imports.xml]: `don't` => `doesn't`
`soring` => `sorting`

File: src/xdocs/config_imports.xml [hunk 7]
```diff
@@ -1318,8 +1319,9 @@ import java.util.stream.IntStream;
           </tr>
           <tr>
             <td>sortStaticImportsAlphabetically</td>
-            <td>whether static imports grouped by <b>top</b> or <b>bottom</b> option
-            are sorted alphabetically or not</td>
+            <td>whether <b>static imports</b> grouped by <b>top</b> or <b>bottom</b> option
+                are sorted alphabetically or not.
+                Attention:it is applied for all static imports as one group.</td>
```
Comment (rnveach) [inline src/xdocs/config_imports.xml]: => `Attention: It is applied to all static imports as one group.`

**Comment by romani** [comment]
@djydewang , fix for shippable is merged to master, so please rebase to avoid this problem in CI.

File: src/xdocs/config_imports.xml (hunk 6 above)
Comment (djydewang) [reply src/xdocs/config_imports.xml]: Done.

File: src/xdocs/config_imports.xml (hunk 7 above)
Comment (djydewang) [reply src/xdocs/config_imports.xml]: Done.

**Comment by djydewang** [comment]
Done.

File: src/xdocs/config_imports.xml [hunk 8]
```diff
@@ -1291,7 +1291,8 @@ import java.util.stream.IntStream;
           </tr>
           <tr>
             <td>ordered</td>
-            <td>whether imports within group should be sorted</td>
+            <td>whether imports within group should be sorted
+              (It don't affect sorting for static imports.)</td>
```
Comment (rnveach) [inline src/xdocs/config_imports.xml]: `don't` => `doesn't`
This was not done from a previous item.

Comment (djydewang) [reply src/xdocs/config_imports.xml]: Done.

**Pull request merged**